include/openthread/srp_client.h

```diff
@@ -0,0 +1,489 @@
+ * and the full list of services is provided as input parameters in the callback. Note that host info and services each
+ * track its own state in the corresponding `mState` member variable of the related data structure (the state
+ * indicating whether the host-info/service is registered or removed or still being added/removed, etc).
+ *
+ * The list of removed services is passed as its own linked-list `aRemovedServices` in the callback. Note that when the
+ * callback is invoked, the SRP client (OpenThread implementation) is done with the removed service instances listed in
+ * `aRemovedServcies` and no longer tracks/stores them (i.e., if from the callback we call `otSrpClientGetServices()`
+ * the removed services will not be present in the returned list). Providing a separate list of removed services in
+ * the callback helps indicate to user which items are now removed and allow user to re-claim/reuse the instances.
+ *
+ * If the server rejects an SRP update request, the DNS repose code (RFC 2136) is mapped to the following errors:
+ *
+ *    (0)  NOERROR   Success (no error condition)                    -> OT_ERROR_NONE
+ *    (1)  FORMERR   Server unable to interpret due to format error  -> OT_ERROR_PARSE
+ *    (2)  SERVFAIL  Server encountered an internal failure          -> OT_ERROR_FAILED
+ *    (3)  NXDOMAIN  Name that ought to exist, does not exists       -> OT_ERROR_NOT_FOUND
+ *    (4)  NOTIMP    Server does not support the query type (OpCode) -> OT_ERROR_NOT_IMPLEMENTED
+ *    (5)  REFUSED   Server refused for policy/security reasons      -> OT_ERROR_SECURITY
+ *    (6)  YXDOMAIN  Some name that ought not to exist, does exist   -> OT_ERROR_DUPLICATED
+ *    (7)  YXRRSET   Some RRset that ought not to exits, does exist  -> OT_ERROR_DUPLICATED
+ *    (8)  NXRRSET   Some RRset that out to exist, does not exist    -> OT_ERROR_NOT_FOUND
+ *    (9)  NOTAUTH   Service is not authoritative for zone           -> OT_ERROR_SECURITY
+ *    (10) NOTZONE   A name is not in the zone                       -> OT_ERROR_PARSE
+ *    (20) BADNAME   Bad name                                        -> OT_ERROR_PARSE
+ *    (21) BADALG    Bad algorithm                                   -> OT_ERROR_SECURITY
+ *    (22) BADTRUN   Bad truncation                                  -> OT_ERROR_PARSE
+ *    Other response codes                                           -> OT_ERROR_FAILED
+ *
+ * The following errors are also possible:
+ *
+ *    OT_ERROR_RESPONSE_TIMEOUT : Timed out waiting for response from server (client would continue to retry).
+ *    OT_ERROR_INVALID_ARGS     : The provided service structure is invalid (e.g., bad service name or `otSrpTxtEntry`).
+ *    OT_ERROR_NO_BUFS          : Insufficient buffer to prepare or send the update message.
+ *
+ * Note that in case of any failure, the client continues the operation, i.e. it prepares and (re)transmits the SRP
```

Comment:
From the client code perspective, it would keep retrying (with an increasing retry wait delay) until it's explicitly stopped.

The idea is that the user would need to decide (probably based on the error) if she/he wants the client to stop or change anything (e.g. if there is a service/host name conflict, the user can remove the conflicting entry and add a new one).

Reply:
I have been looking into different options related to this. I feel it gets rather complicated (what to remove and on which errors and how to report it to user) and it adds more code for handling these corner-cases. 

I suggest for now we keep it simple, i.e. clinet keep retrying (with an increasting rety wait interval) until it is stopped. And rely on the user to handle errors (bad name, etc).